Implemented Reading List calls

luca-schlecker opened the pull request:
Closes #172
Closes #167
Closes #173
Closes #174

## What I have made

- Add PaperTableRepo + Tests
- Add AuthorTableRepo + Tests
- Add ReadingListTableRepo + Tests
- Add ReadingListService

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does
not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] (for reviewer) I have checked the implementation against the requirements

Slartibartfass2 commented:
In GitLab by @dts84 on Apr 27, 2022, 20:21

mentioned in commit 4f85da2a57d8399d29c19f0e7f1c7b78ece48a42

Merseleo commented:
Maybe we should wait until there really exists paper in the database, right? Because otherwise it is quite complicate to test the reading list properly. 

And one more request: use also implemented the service functions to check for existing on reading list, add and remove paper to / from reading list as well. As I skimmed through I noticed that these functions are not exactly implemented as specified in the API documentation and you should link there issues (#167, #173, #174) and adapt the title of the PR.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/PaperTableRepo.kt:
```diff
@@ -0,0 +1,45 @@
+import se.uulm.snowballr.backend.table.toPaper
+import java.util.UUID
+
+interface IPaperTableRepo {
```
Add a comment for this interface similar to the other repo interfaces.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/PaperTableRepo.kt:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Defines an interface for repository operations related to the [PaperTable].
+ *
+ * This interface is used to handle persistence and retrieval operations for papers by providing
```
This is not an interface.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/PaperTableRepo.kt:
```diff
@@ -0,0 +1,45 @@
+package se.uulm.snowballr.backend.repository
+
+import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
+import org.jetbrains.exposed.sql.selectAll
```
Unused import

```suggestion
import org.jetbrains.exposed.sql.selectAll
```

luca-schlecker replied:
See [my previous comment](https://github.com/SE-UUlm/snowballr-backend/pull/223/files#r2231911836) for my response.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/PaperTableRepoTest.kt:
```diff
@@ -0,0 +1,68 @@
+            val paperId = insertTestPaperAndGetId()
+            val paper = repo.getPaperById(paperId)
+
+            with(paper) {
```
Interesting syntax; didn't know this.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/AuthorTableRepo.kt:
```diff
@@ -0,0 +1,45 @@
+    suspend fun getAuthorById(id: UUID): Author
+}
+
+class AuthorTableRepo(
```
Add a comment for this repo similar to the other repos.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/AuthorTableRepo.kt:
```diff
@@ -0,0 +1,45 @@
+package se.uulm.snowballr.backend.repository
+
+import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
+import org.jetbrains.exposed.sql.selectAll
```
Unused import

```suggestion
import org.jetbrains.exposed.sql.selectAll
```

luca-schlecker replied:
See [my previous comment](https://github.com/SE-UUlm/snowballr-backend/pull/223/files#r2231911836) for my response.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepo.kt:
```diff
@@ -0,0 +1,100 @@
+    suspend fun getAllReadingListEntries(userId: UUID): List<UUID>
+}
+
+class ReadingListTableRepo(
```
Add a comment for this repo similar to the other repos.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepo.kt:
```diff
@@ -0,0 +1,100 @@
+        if (ReadingListTable.selectAll().where {
+                (ReadingListTable.userId eq userId) and (ReadingListTable.paperId eq paperId)
+            }.count() > 0
+        ) {
```
Here we can reuse `isPaperOnReadingList`
```suggestion
        if (isPaperOnReadingList(userId, paperId)) {
```

luca-schlecker replied:
AFAIK nested transactions are ok, adding this to improve readability.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepo.kt:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    override suspend fun isPaperOnReadingList(userId: UUID, paperId: UUID): Boolean = db.dbQuery {
+        ReadingListTable.selectAll().where {
```
Do you know if it makes a difference if we only select the ID, for example?

luca-schlecker replied:
In SQL I'd prefer `COUNT(*)`. The NULL-behavior is different. AFAIK `COUNT(*)` is usually preferred for counting rows.

Slartibartfass2 replied:
I see. This call currently resolves to
`SELECT reading_list.paper_id, reading_list.user_id, reading_list.created_at FROM reading_list WHERE (reading_list.user_id = '...') AND (reading_list.paper_id = '...') LIMIT 1`
and the rest is done on the client side.
Using `count()` at the end resolves to
`SELECT COUNT(*) FROM reading_list WHERE (reading_list.user_id = '...') AND (reading_list.paper_id = '...')`.
I did a bit of research, and according to that, the latter is better since we don't retrieve the data (although it's just one entry) but just count in the DB.

I keep this up to you if we still should use `count()` instead.

luca-schlecker replied:
I'd prefer the latter then and will restore the count version.

Slartibartfass2 replied:
Great

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepo.kt:
```diff
@@ -0,0 +1,100 @@
+    suspend fun isPaperOnReadingList(userId: UUID, paperId: UUID): Boolean
+
+    /**
+     * @returns The [List] of all ids of the papers the provided user has on their reading list.
```
```suggestion
     * @returns The [List] of all IDs of the papers the provided user has on their reading list.
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepo.kt:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    private fun throwIfPaperDoesNotExist(paperId: UUID) {
+        if (PaperTable.selectAll().where { PaperTable.id eq paperId }.count() == 0L) {
```
Extract these two queries to separate calls or variables for improved readability.

Slartibartfass2 replied:
Also, we can use `empty()` instead of `count() == 0L`, which is more descriptive.

Slartibartfass2 replied:
Also, so far we have moved such logic to the business logic of the service layer to keep the repository layer as simple as possible. There, you can use the `getUserById` to check for a user's existence.

luca-schlecker replied:
> Extract these two queries to separate calls or variables for improved readability.

I do not quite understand what you are trying to tell me.

luca-schlecker replied:
> Also, we can use `empty()` instead of `count() == 0L`, which is more descriptive.

Implemented this.

luca-schlecker replied:
> Also, so far we have moved such logic to the business logic of the service layer to keep the repository layer as simple as possible. There, you can use the `getUserById` to check for a user's existence.

Shouldn't such checks be wrapped in a single transaction together with the rest of the db code, though?

Slartibartfass2 replied:
Why should they? Since we only read from the DB this isn't a problem.

The current service implementation already checks whether the user exists, so there's no need to check it another time.
When also checking the paper's existence we can remove the checks in the repo.
```kotlin
    override suspend fun removePaperFromReadingList(paperId: Base.Id): Base.Nothing {
        val currentUser = userRepo.getUserById(GrpcContext.getUserIdFromContext())
        val paper = paperRepo.getPaperById(parseUUID(paperId.id, EntityType.PAPER))
        repo.removeReadingListEntry(currentUser.id, paper.id)
        return Base.Nothing.newBuilder().build()
    }
```
and
```kotlin
    override suspend fun createReadingListEntry(userId: UUID, paperId: UUID) = db.dbQuery {
        if (isPaperOnReadingList(userId, paperId)) {
            // Entry already exists, so we do nothing to avoid duplicates.
            return@dbQuery
        }

        ReadingListTable.insert {
            it[this.userId] = userId
            it[this.paperId] = paperId
        }
        Unit
    }
```

luca-schlecker replied:
> ```kotlin
>     override suspend fun removePaperFromReadingList(paperId: Base.Id): Base.Nothing {
>         val currentUser = userRepo.getUserById(GrpcContext.getUserIdFromContext())
>         val paper = paperRepo.getPaperById(parseUUID(paperId.id, EntityType.PAPER))
>         repo.removeReadingListEntry(currentUser.id, paper.id)
>         return Base.Nothing.newBuilder().build()
>     }
> ```

Why get the paper using the ID first just to only use the ID then? Seems unnecessary

Slartibartfass2 replied:
What about `doesPaperExistById`?

Slartibartfass2 replied:
The logic for testing whether a paper exists is implemented three times currently (`CitationTableRepo`, `AuthorOfPaperTableRepo` and `ReadingListTableRepo`). This should give us enough reason to make this a separate call in the `PaperTable` and use it on the service layer.

luca-schlecker replied:
I moved the `throwIfPaperDoesNotExist` call into the `PaperTable` and also extracted a `doesPaperExistById` function which is also accessible in the `PaperTable`. The separate implementations in the mentioned files have been removed.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepo.kt:
```diff
@@ -0,0 +1,100 @@
+import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
+import org.jetbrains.exposed.sql.and
+import org.jetbrains.exposed.sql.count
+import org.jetbrains.exposed.sql.deleteWhere
```
Unused import
```suggestion
import org.jetbrains.exposed.sql.deleteWhere
```

luca-schlecker replied:
The `count` function is used in line 50. Yet, it compiles without the import. I believe there is something wrong with the linter, as it does not flag this.

I rely on the `lint` and `format` tasks of gradle to handle this and they don't seem to mind this import. How should this be handled in the future?

Slartibartfass2 replied:
I once read about detekt having problems with some false negatives, but they wouldn't improve the analysis as it wouldn't be worth the performance decrease. I can't find the issue/PR now. Since they still detect most of the unused imports, this is fine, and unused imports don't hurt. IntelliJ does seem to have a better analysis for this and detects those false negatives.

I think it's fine when they are accidentally added, but when one of us notices them, we can remove them. So in this case, I would just remove them.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepoTest.kt:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    @Suppress("LongParameterList")
+    private suspend fun insertTestUserAndGetId(
```
Use `createExampleUser` from `RepositoryHelper` instead.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepo.kt:
```diff
@@ -0,0 +1,100 @@
+    suspend fun removeReadingListEntry(userId: UUID, paperId: UUID)
+
+    /**
+     * @returns Wether the provided paper is on the reading list of the provided user.
```
```suggestion
     * @returns Whether the provided paper is on the reading list of the provided user.
```

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepoTest.kt:
```diff
@@ -0,0 +1,224 @@
+    @Nested
+    inner class GetAllReadingListEntries {
+        @Test
+        fun `When there are no papers on a users reading list, then getAllReadingListEntries returns an empty list`() =
```
```suggestion
        fun `When there are no papers on a user's reading list, then getAllReadingListEntries returns an empty list`() =
```

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepoTest.kt:
```diff
@@ -0,0 +1,224 @@
+            }
+
+        @Test
+        fun `When there is one paper on a users reading list, then getAllReadingListEntries returns a list with that paper`() =
```
```suggestion
        fun `When there is one paper on a user's reading list, then getAllReadingListEntries returns a list with that paper`() =
```

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepoTest.kt:
```diff
@@ -0,0 +1,224 @@
+            }
+
+        @Test
+        fun `When there are two papers on a users reading list, then getAllReadingListEntries returns a list with those two papers`() =
```
```suggestion
        fun `When there are two papers on a user's reading list, then getAllReadingListEntries returns a list with those two papers`() =
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepo.kt:
```diff
@@ -0,0 +1,100 @@
+    override suspend fun isPaperOnReadingList(userId: UUID, paperId: UUID): Boolean = db.dbQuery {
+        ReadingListTable.selectAll().where {
+            (ReadingListTable.userId eq userId) and (ReadingListTable.paperId eq paperId)
+        }.count() > 0
```
```suggestion
        }.empty().not()
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/ReadingListService.kt:
```diff
@@ -0,0 +1,80 @@
+
+/**
+ * Default implementation of [IReadingListService].
+ */
```
Provide further information on this service similar to other services.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepo.kt:
```diff
@@ -0,0 +1,100 @@
+        }.count() > 0
+    }
+
+    override suspend fun getAllReadingListEntries(userId: UUID): List<UUID> = db.dbQuery {
```
For the use in the service layer, we could make this query use a join so that we get the paper list directly.

Slartibartfass2 reviewed (changes requested):
Edit: Please add your changes as new commits so that I can see the diff. Also, don't rebase until we're done here. Before this is merged, you can squash the commits so that a clean history is created.

The calls are straightforward; good job implementing them.

The service calls are neither tested nor connected to the main service. This gives me reason to believe that you didn't even test those calls yourselves, even though we agreed on doing that in one of our last meetings. Please do so in the future starting with this PR.

When I start the backend, I receive the following exception:
```kotlin
[25-07-25 16:43:19.883] DEBUG [main] Exposed: CREATE EXTENSION IF NOT EXISTS hstore;
[25-07-25 16:43:19.883] WARN  [main] Exposed: Transaction attempt #0 failed: org.postgresql.util.PSQLException: ERROR: no schema has been selected to create in. Statement(s): CREATE EXTENSION IF NOT EXISTS hstore;
org.jetbrains.exposed.exceptions.ExposedSQLException: org.postgresql.util.PSQLException: ERROR: no schema has been selected to create in
        at org.jetbrains.exposed.sql.statements.Statement.executeIn$exposed_core(Statement.kt:99)
...
```
This can be fixed when moving the line `exec("CREATE EXTENSION IF NOT EXISTS hstore;")` in `Database.kt` after the line `SchemaUtils.setSchema(schema)`. Please fix this.

As @Merseleo already mentioned, mark the other issues as closed by this PR and update the PR title.
For testing, please add some papers on startup and make this a test branch so we can thoroughly test this behavior.

luca-schlecker reviewed (commented):
Seems like I was very overzealous when I marked this PR as reviewable, sorry for that. This was a disaster. But, everything should be fixed now. I've also manually validated the functionality using a test branch: `test/172-implement-getreadinglist-call`.

The gradle linter/formatter tasks seems to struggle with the imports you've flagged, we should find a solution to that as I (and the pipeline) rely on them.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepo.kt:
```diff
@@ -0,0 +1,108 @@
+ * facilitates CRUD operations on project member records associated with a given project and
+ * ensures database transactions are handled properly.
+ *
+ * @param db The database abstraction used for executing queries within a transaction.
```
This is the documentation for the `ProjectMemberTableRepo`, but not the `ReadingListTableRepo`.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/AuthorOfPaperTableRepo.kt:
```diff
@@ -0,0 +1,58 @@
+        if (PaperTable.selectAll().where { PaperTable.id eq paperId }.empty()) {
+            throw NotFoundException(EntityType.PAPER, paperId.toString())
+        }
+    }
```
As already commented in the ReadingListRepo, this logic should be in the service layer.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/AuthorOfPaperTableRepo.kt:
```diff
@@ -0,0 +1,58 @@
+    /**
+     * Returns all authors of a [Paper] or throws a [NotFoundException] if the paper with the passed [paperId] doesn't exist.
+     *
+     * @param paperId The id of the paper for which the authors should be retrieved.
```
```suggestion
     * @param paperId The ID of the paper for which the authors should be retrieved.
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/AuthorOfPaperTableRepo.kt:
```diff
@@ -0,0 +1,58 @@
+package se.uulm.snowballr.backend.repository.association
+
+import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
+import org.jetbrains.exposed.sql.selectAll
```
Unused import.
```suggestion
import org.jetbrains.exposed.sql.selectAll
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/AuthorOfPaperTableRepo.kt:
```diff
@@ -0,0 +1,58 @@
+ */
+interface IAuthorOfPaperTableRepo {
+    /**
+     * Returns all authors of a [Paper] or throws a [NotFoundException] if the paper with the passed [paperId] doesn't exist.
```
This exceeds the maximum line length of 120. (we have no linter for comment line lengths)
```suggestion
     * Returns all authors of a [Paper] or throws a [NotFoundException] if the paper with the passed [paperId] doesn't
     * exist.
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/CitationTableRepo.kt:
```diff
@@ -0,0 +1,41 @@
+package se.uulm.snowballr.backend.repository.association
+
+import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
+import se.uulm.snowballr.backend.db.IDatabase
```
Unused import.
```suggestion
import se.uulm.snowballr.backend.db.IDatabase
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/CitationTableRepo.kt:
```diff
@@ -0,0 +1,41 @@
+ * Defines an interface for repository operations related to the [CitationTable].
+ *
+ * This interface provides abstraction for handling persistence and retrieval operations for citations of papers. By using this
+ * interface, the functionality for managing papers can remain decoupled from the specifics of the database layer.
```
Exceeds the maximum line length.
```suggestion
 * This interface provides abstraction for handling persistence and retrieval operations for citations of papers. By
 * using this interface, the functionality for managing papers can remain decoupled from the specifics of the database
 * layer.
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/CitationTableRepo.kt:
```diff
@@ -0,0 +1,41 @@
+ */
+interface ICitationTableRepo {
+    /**
+     * Returns the backward references of a paper by its ID or throws a [NotFoundException] if the paper with the passed [id] doesn't exist.
```
Exceeds the maximum line length.
```suggestion
     * Returns the backward references of a paper by its ID or throws a [NotFoundException] if the paper with the passed
     * [id] doesn't exist.
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/CitationTableRepo.kt:
```diff
@@ -0,0 +1,41 @@
+ *
+ * This class provides functionality to handle persistence and retrieval operations for citations of papers by leveraging the
+ * database abstraction defined in [IDatabase]. It facilitates CRUD operations on citations of papers and ensures
+ * database transactions are handled properly.
```
Exceeds the maximum line length.
```suggestion
 * This class provides functionality to handle persistence and retrieval operations for citations of papers by
 * leveraging the database abstraction defined in [IDatabase]. It facilitates CRUD operations on citations of papers and
 * ensures database transactions are handled properly.
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/CitationTableRepo.kt:
```diff
@@ -0,0 +1,41 @@
+class CitationTableRepo(
+    private val db: IDatabase,
+) : ICitationTableRepo {
+    override suspend fun getBackwardsReferencedPaperIdsOfPaperById(id: UUID): List<UUID> = db.dbQuery {
```
The documentation comment says it throws a `NotFoundException`, but it doesn't.

luca-schlecker replied:
Now throws an according exception

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/association/CitationTableRepoTest.kt:
```diff
@@ -0,0 +1,82 @@
+            it[PaperTable.publicationName] = publicationName
+            it[PaperTable.fetcherMetadata] = fetcherMetadata
+        }.value
+    }
```
We added this test helper in four different test files. What about moving this to `RepositoryHelper` and then using this single implementation for all test files?

Slartibartfass2 replied:
It's still duplicated in `PaperTableRepoTest`.

luca-schlecker replied:
Ah yes, it evaded my search as it had a different name. Removed that now.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/association/CitationTableRepoTest.kt:
```diff
@@ -0,0 +1,82 @@
+        @Test
+        fun `When a paper has one backward reference, then it is returned`() = runTest {
+            val paperId = insertPaperAndGetId()
+            val paperId2 = insertPaperAndGetId(externalId = "ExternalId2")
+            addCitation(paperId2, paperId)
+            val actualReferences = repo.getBackwardsReferencedPaperIdsOfPaperById(paperId)
+            assertThat(actualReferences).hasSize(1)
+            assertThat(actualReferences).containsExactly(paperId2)
+        }
+
+        @Test
+        fun `When a paper has two backward references, then they are returned`() = runTest {
+            val paperId = insertPaperAndGetId()
+            val paperId2 = insertPaperAndGetId(externalId = "ExternalId2")
+            val paperId3 = insertPaperAndGetId(externalId = "ExternalId3")
```
Any reason why the other papers have an external ID, but we don't test for those?

luca-schlecker replied:
ExternalID has a unique constraint in the DB, requiring the second and third paper to choose something different.
Furthermore, the external ids are compared in the last assertion in that function.

Slartibartfass2 replied:
I see.
However, I can't see where the external IDs are asserted. Can you tell me the line?

luca-schlecker replied:
```kotlin
assertThat(actualReferences).containsExactlyInAnyOrder(paperId2, paperId3)
```

In (currently) line 51. The whole Paper is compared and thus that `externalId` is implicitly compared with it.

Slartibartfass2 replied:
Ah, I see, I thought you meant that they were explicitly compared.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/ReadingListService.kt:
```diff
@@ -0,0 +1,94 @@
+    private val citationRepo: ICitationTableRepo,
+    private val repo: IReadingListTableRepo,
+) : IReadingListService {
+    override suspend fun getReadingList(): PaperOuterClass.Paper.List {
```
Currently, when I run this with a paper on the reading list, the following exception is thrown:
```
[25-07-27 19:02:15.317] DEBUG [DefaultDispatcher-worker-4] Exposed: SELECT reading_list.paper_id, reading_list.user_id, reading_list.created_at, paper.id, paper.title, paper.external_id, paper.abstract, paper.published_at, paper.publisher, paper.publication_type, paper.publication_name, paper.pdf_id, paper.fetcher_metadata, paper.created_at, paper.modified_at, paper.modified_by FROM reading_list INNER JOIN paper ON paper.id = reading_list.paper_id WHERE reading_list.user_id = 'bd47f937-604a-48c3-83d3-804a94c775d3'
[25-07-27 19:02:15.329] ERROR [DefaultDispatcher-worker-1] s.u.s.b.g.i.ExceptionInterceptor: gRPC call failed due to an unexpected exception
java.lang.StringIndexOutOfBoundsException: Index 5 out of bounds for length 2
        at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:55)
        at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:52)
        at java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:213)
        at java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:210)
        at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:98)
        at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
        at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
        at java.base/java.lang.String.checkIndex(String.java:4832)
        at java.base/java.lang.StringLatin1.charAt(StringLatin1.java:46)
        at java.base/java.lang.String.charAt(String.java:1555)
        at org.postgresql.util.HStoreConverter.fromString(HStoreConverter.java:132)
        at se.uulm.snowballr.backend.table.ColumnHelperKt.stringMap$lambda$3(ColumnHelper.kt:63)
        at org.jetbrains.exposed.sql.ColumnTypeKt$columnTransformer$1.wrap(ColumnType.kt:297)
        at org.jetbrains.exposed.sql.ColumnWithTransform.valueFromDB(ColumnType.kt:361)
        at org.jetbrains.exposed.sql.ResultRow.rawToColumnValue(ResultRow.kt:97)
        at org.jetbrains.exposed.sql.ResultRow.getInternal$lambda$5$lambda$4$lambda$3(ResultRow.kt:85)
        at org.jetbrains.exposed.sql.vendors.DatabaseDialectKt.withDialect(DatabaseDialect.kt:235)
        at org.jetbrains.exposed.sql.ResultRow.getInternal$lambda$5(ResultRow.kt:84)
        at org.jetbrains.exposed.sql.ResultRow$ResultRowCache.cached(ResultRow.kt:216)
        at org.jetbrains.exposed.sql.ResultRow.getInternal(ResultRow.kt:71)
        at org.jetbrains.exposed.sql.ResultRow.get(ResultRow.kt:40)
        at se.uulm.snowballr.backend.table.PaperTableKt.toPaper(PaperTable.kt:65)
        at se.uulm.snowballr.backend.repository.association.ReadingListTableRepo$getAllReadingListEntries$2.invokeSuspend(ReadingListTableRepo.kt:105)
        at se.uulm.snowballr.backend.repository.association.ReadingListTableRepo$getAllReadingListEntries$2.invoke(ReadingListTableRepo.kt)
        at se.uulm.snowballr.backend.repository.association.ReadingListTableRepo$getAllReadingListEntries$2.invoke(ReadingListTableRepo.kt)
        at se.uulm.snowballr.backend.db.Database$dbQuery$2.invokeSuspend(Database.kt:126)
        at se.uulm.snowballr.backend.db.Database$dbQuery$2.invoke(Database.kt)
        at se.uulm.snowballr.backend.db.Database$dbQuery$2.invoke(Database.kt)
        at org.jetbrains.exposed.sql.transactions.experimental.SuspendedKt$suspendedTransactionAsyncInternal$1.invokeSuspend(Suspended.kt:193)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
        at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:113)
        at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:89)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:586)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:820)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:717)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:704)
```

luca-schlecker replied:
The issue is that the `HStoreConverter` does not correctly parse "{}". Trimming these brackets from the string solves this problem, only to reveal a current issue with the `Paper.toGrpcPaper()` method. The data class allows fields to be NULL, just as the database, but the `set...` methods of the gRPC message builder does not. This has gone unnoticed until now. Seems like you overlooked some tests and docs there too. 😉

Slartibartfass2 replied:
Alright, good job debugging this. Luckily this would've been noticed when service tests were added.

luca-schlecker replied:
We've got there in the end. This issue seems to be resolved then, or is something still missing?

Slartibartfass2 reviewed (changes requested):
Good job. We made some improvements. Still no tests for the `ReadingListService` exist.

The testing branch is great, thanks for adding it.

Slartibartfass2 reviewed (changes requested):
Please provide tests for the service layer before we proceed any further.

Slartibartfass2 reviewed (changes requested):
Please read [our testing guidelines](https://github.com/SE-UUlm/snowballr-backend/wiki/Testing#service) and refactor the service tests accordingly.

luca-schlecker commented:
> Please read [our testing guidelines](https://github.com/SE-UUlm/snowballr-backend/wiki/Testing#service) and refactor the service tests accordingly.

I do not think splitting this 100 line test into four files serves our project. The testing guideline talks about service methods being _"expected to contain a lot of test cases"_. As the main logic is contained in the `ReadingListTableRepo` anyway, the tests for the `ReadingListService` remain few. Furthermore, I do not expect there to be additional methods for the `ReadingListService` as the use-case for reading-lists is completed. And even if, a refactoring-effort can then still be made.

Slartibartfass2 commented:
> I do not think splitting this 100 line test into four files serves our project. The testing guideline talks about service methods being _"expected to contain a lot of test cases"_. As the main logic is contained in the `ReadingListTableRepo` anyway, the tests for the `ReadingListService` remain few. Furthermore, I do not expect there to be additional methods for the `ReadingListService` as the use-case for reading-lists is completed. And even if, a refactoring-effort can then still be made.

I think it does serve our project since (1.) we've done this consistently so far, and this would break the consistency, and more important (2.) splitting them up makes it easier to compare the implementation with the tests in order to see which branches are already tested and which still require testing.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/association/AuthorOfPaperTableRepoTest.kt:
```diff
@@ -0,0 +1,76 @@
+import java.util.UUID
+
+@ExperimentalCoroutinesApi
+@DelicateCoroutinesApi
```
This can be removed.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/association/CitationTableRepoTest.kt:
```diff
@@ -0,0 +1,58 @@
+import java.util.UUID
+
+@ExperimentalCoroutinesApi
+@DelicateCoroutinesApi
```
This can be removed.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepoTest.kt:
```diff
@@ -0,0 +1,185 @@
+import java.util.UUID
+
+@ExperimentalCoroutinesApi
+@DelicateCoroutinesApi
```
This can be removed.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/AuthorTableRepoTest.kt:
```diff
@@ -0,0 +1,53 @@
+import java.util.UUID
+
+@ExperimentalCoroutinesApi
+@DelicateCoroutinesApi
```
This can be removed.

Slartibartfass2 replied:
This is still there.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/PaperTableRepoTest.kt:
```diff
@@ -0,0 +1,68 @@
+import java.util.UUID
+
+@ExperimentalCoroutinesApi
+@DelicateCoroutinesApi
```
This can be removed.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/readinglist/AddPaperToReadingListTest.kt:
```diff
@@ -0,0 +1,42 @@
+import java.util.UUID
+
+@DelicateCoroutinesApi
+@ExperimentalCoroutinesApi
```
This can be removed.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/readinglist/GetReadingListTest.kt:
```diff
@@ -0,0 +1,49 @@
+import se.uulm.snowballr.backend.service.MainServiceTest
+
+@DelicateCoroutinesApi
+@ExperimentalCoroutinesApi
```
This can be removed.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/readinglist/IsPaperOnReadingListTest.kt:
```diff
@@ -0,0 +1,50 @@
+import kotlin.test.assertTrue
+
+@DelicateCoroutinesApi
+@ExperimentalCoroutinesApi
```
This can be removed.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/readinglist/RemovePaperFromReadingListTest.kt:
```diff
@@ -0,0 +1,42 @@
+import java.util.UUID
+
+@DelicateCoroutinesApi
+@ExperimentalCoroutinesApi
```
This can be removed.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/readinglist/IsPaperOnReadingListTest.kt:
```diff
@@ -0,0 +1,50 @@
+            coVerify(exactly = 1) { readingListRepoMock.isPaperOnReadingList(user.id, paper2.id) }
+        }
+
+    fun UUID.toGrpcId() = Base.Id.newBuilder().setId(this.toString()).build()
```
This is used in three files. Should we extract it?

luca-schlecker replied:
I've put it in the `DataBuilder`. 👍

Slartibartfass2 replied:
It's still there and nothing in `DataBuilder`.

luca-schlecker replied:
Weird. Now it should be there.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/readinglist/AddPaperToReadingListTest.kt:
```diff
@@ -0,0 +1,42 @@
+
+@DelicateCoroutinesApi
+@ExperimentalCoroutinesApi
+class AddPaperToReadingListTest : MainServiceTest() {
```
Add a test case where `repo.createReadingListEntry(...)` throws.

Slartibartfass2 replied:
This is missing.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/readinglist/GetReadingListTest.kt:
```diff
@@ -0,0 +1,49 @@
+
+@DelicateCoroutinesApi
+@ExperimentalCoroutinesApi
+class GetReadingListTest : MainServiceTest() {
```
Add a test case where `repo.getAllReadingListEntries(...)` throws.

luca-schlecker replied:
There should already be one.

Slartibartfass2 replied:
There are only two tests, in one `GrpcContext.getUserIdFromContext()` throws, and in the other one the call is successful.

So this is missing.

luca-schlecker replied:
Now I get what you're trying to tell me. I thought we assumed the repositories as infallible, but I checked out some other service tests and see what you mean. Added this.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/readinglist/IsPaperOnReadingListTest.kt:
```diff
@@ -0,0 +1,50 @@
+
+@DelicateCoroutinesApi
+@ExperimentalCoroutinesApi
+class IsPaperOnReadingListTest : MainServiceTest() {
```
Add a test case where `repo.isPaperOnReadingList(...)` throws.

Slartibartfass2 replied:
This is missing.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/readinglist/RemovePaperFromReadingListTest.kt:
```diff
@@ -0,0 +1,42 @@
+
+@DelicateCoroutinesApi
+@ExperimentalCoroutinesApi
+class RemovePaperFromReadingListTest : MainServiceTest() {
```
Add a test case where `repo.removeReadingListEntry(...)` throws.

Slartibartfass2 replied:
This is missing.

Slartibartfass2 reviewed (changes requested):
Good job. The service tests look good. I like that you use verify, which we could probably use more often.

Furthermore, I think we should rebase this, since a lot happened on `develop` (see the "This can be removed." comments).

After that we're nearly done.

luca-schlecker committed:
feat: implement reading lists

luca-schlecker commented:
This branched is now fully rebased on `develop`. I extracted functions, added error-path tests and should also have addressed every comment of yours. Furthermore, the commits are now squashed and align with the conventional commit requirement.

During this rebase, I've encountered issues with the new test-setup, though:
<details>
  <summary>Error Log</summary>
  <pre><code>org.jetbrains.exposed.exceptions.ExposedSQLException: org.postgresql.util.PSQLException: ERROR: cache lookup failed for type 16386
SQL: [INSERT INTO paper (id, created_at, title, external_id, abstract, published_at, publisher, publication_type, publication_name, fetcher_metadata) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)]
    at app//org.jetbrains.exposed.sql.statements.Statement.executeIn$exposed_core(Statement.kt:99)
    at app//org.jetbrains.exposed.sql.Transaction.exec(Transaction.kt:292)
    at app//org.jetbrains.exposed.sql.Transaction.exec(Transaction.kt:269)
    at app//org.jetbrains.exposed.sql.statements.Statement.execute(Statement.kt:59)
    at app//se.uulm.snowballr.backend.repository.RepositoryHelper$insertPaperAndGetId$2.invokeSuspend(RepositoryHelper.kt:92)
    at app//se.uulm.snowballr.backend.repository.RepositoryHelper$insertPaperAndGetId$2.invoke(RepositoryHelper.kt)
    at app//se.uulm.snowballr.backend.repository.RepositoryHelper$insertPaperAndGetId$2.invoke(RepositoryHelper.kt)
    at app//se.uulm.snowballr.backend.repository.RepositoryTest$TestDatabase$query$2.invokeSuspend(RepositoryTest.kt:93)
    at app//se.uulm.snowballr.backend.repository.RepositoryTest$TestDatabase$query$2.invoke(RepositoryTest.kt)
    at app//se.uulm.snowballr.backend.repository.RepositoryTest$TestDatabase$query$2.invoke(RepositoryTest.kt)
    at app//org.jetbrains.exposed.sql.transactions.experimental.SuspendedKt$suspendedTransactionAsyncInternal$1.invokeSuspend(Suspended.kt:193)
    at app//kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at app//kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:100)
    at app//kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:124)
    at app//kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:89)
    at app//kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:586)
    at app//kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:820)
    at app//kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:717)
    at app//kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:704)
Caused by: org.postgresql.util.PSQLException: ERROR: cache lookup failed for type 16386
    at app//org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2734)
    at app//org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2421)
    at app//org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:372)
    at app//org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:518)
    at app//org.postgresql.jdbc.PgStatement.execute(PgStatement.java:435)
    at app//org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:196)
    at app//org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:157)
    at app//com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
    at app//com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
    at app//org.jetbrains.exposed.sql.statements.jdbc.JdbcPreparedStatementImpl.executeUpdate(JdbcPreparedStatementImpl.kt:53)
    at app//org.jetbrains.exposed.sql.statements.InsertStatement.execInsertFunction(InsertStatement.kt:232)
    at app//org.jetbrains.exposed.sql.statements.InsertStatement.executeInternal(InsertStatement.kt:244)
    at app//org.jetbrains.exposed.sql.statements.InsertStatement.executeInternal(InsertStatement.kt:19)
    at app//org.jetbrains.exposed.sql.statements.Statement.executeIn$exposed_core(Statement.kt:97)
    ... 18 more</code></pre>
</details>

I've been able to work around them by commenting `removeExtensions()` in `RepositoryTest.kt#L147`. I can only speculate why this may cause the issue. However, I've confirmed this on two machines of mine. The workaround is in a separate commit. Please let me know if you can replicate it and how we should proceed.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/AuthorOfPaperTableRepo.kt:
```diff
@@ -0,0 +1,49 @@
+ */
+interface IAuthorOfPaperTableRepo {
+    /**
+     * Returns all authors of a [Paper] or throws a [NotFoundException] if the paper with the passed [paperId] doesn't
```
This doesn't throw anymore. Update the comment.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/CitationTableRepo.kt:
```diff
@@ -0,0 +1,42 @@
+ */
+interface ICitationTableRepo {
+    /**
+     * Returns the backward references of a paper by its ID or throws a [NotFoundException] if the paper with the passed
```
This doesn't throw anymore. Update the comment.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepo.kt:
```diff
@@ -0,0 +1,89 @@
+) : IReadingListTableRepo {
+    override suspend fun createReadingListEntry(userId: UUID, paperId: UUID) = db.query {
+        if (isPaperOnReadingList(userId, paperId)) {
+            // Entry already exists, so we do nothing to avoid duplicates.
```
Note that this isn't about avoiding duplicates, but rather preventing the DB query to fail because the primary key is not unique.

You can close this yourself.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepo.kt:
```diff
@@ -0,0 +1,89 @@
+            .where { ReadingListTable.userId eq userId }
+            .map { it.toPaper() }
+            .toList()
+    }
```
`.map` already returns a list.
```suggestion
    }
```

luca-schlecker replied:
Oops, still having iterators occupying my mind.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/AuthorTableRepo.kt:
```diff
@@ -0,0 +1,53 @@
+        .selectAll()
+        .where { AuthorTable.id eq id }
+        .map { it.toAuthor() }
+        .singleOrNull()
```
Here you can use one of the new helper methods.

```suggestion
    private fun getAuthorByIdOrNull(id: UUID): Author? = AuthorTable.getEntityByIdOrNull(id, ResultRow::toAuthor)
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/association/ReadingListTableRepo.kt:
```diff
@@ -0,0 +1,89 @@
+    suspend fun isPaperOnReadingList(userId: UUID, paperId: UUID): Boolean
+
+    /**
+     * @returns The [List] of all [Paper]s the provided user has on their reading list.
```
`@returns` doesn't exist in KDoc.
```suggestion
     * @return Whether the provided paper is on the reading list of the provided user.
     */
    suspend fun isPaperOnReadingList(userId: UUID, paperId: UUID): Boolean

    /**
     * @return The [List] of all [Paper]s the provided user has on their reading list.
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/PaperTableRepo.kt:
```diff
@@ -0,0 +1,52 @@
+    suspend fun getPaperById(id: UUID): Paper
+
+    /**
+     * @returns whether the paper with the passed [id] exists.
```
`@returns` doesn't exist in KDoc. There's another occurrence in `FetcherManager`.
```suggestion
     * @return Whether the paper with the passed [id] exists.
```

luca-schlecker replied:
Removed all occurrences I could find :+1:

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/ReadingListService.kt:
```diff
@@ -0,0 +1,116 @@
+    /**
+     * Throws a [NotFoundException] if the paper with the passed [id] doesn't exist.
+     */
+    suspend fun throwIfPaperDoesNotExist(id: UUID) {
```
This doesn't need to be public.
```suggestion
    private suspend fun throwIfPaperDoesNotExist(id: UUID) {
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/ReadingListService.kt:
```diff
@@ -0,0 +1,116 @@
+
+    override suspend fun getReadingList(): PaperOuterClass.Paper.List {
+        val currentUser = userRepo.getUserById(GrpcContext.getUserIdFromContext())
+        val papers = repo.getAllReadingListEntries(currentUser.id).map { paper ->
+            val authors = authorOfPaperRepo.getAuthorsOfPaperById(paper.id).map { it.toGrpcAuthor() }
+            val backwardReferences = citationRepo.getBackwardsReferencedPaperIdsOfPaperById(
+                paper.id,
+            ).map { it.toString() }
+            paper.toGrpcPaper(authors, backwardReferences)
+        }
```
We could probably make this a join later, what are your thoughts?

luca-schlecker replied:
I do not yet see how you would want to go about this. I see no issue with the current solution.

Slartibartfass2 replied:
I'll have to look another time, but this isn't urgent now. No need for premature optimization.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/model/dto/Paper.kt:
```diff
@@ -28,17 +28,22 @@ data class Paper(
+        setTitle(paper.title)
+        paper.externalId?.let { setExternalId(it) }
+        setAbstrakt(paper.abstract)
+        paper.publisher?.let { setPublisher(it) }
```
Similar to this, we have to do it for the ORCID in the author DTO.

Slartibartfass2 reviewed (changes requested):
It appears that some changes have gone missing. Please add them. Otherwise, we're almost done. Good job.

Please don't rebase now. There shouldn't be any conflicts.

luca-schlecker commented:
I could only confirm one missing instance of the
```
@DelicateCoroutinesApi
@ExperimentalCoroutinesApi
```

annotations. The rest were definitely removed previously. The review system may have showed you an older version.
Either way, there is no trace of them in the source code anymore. The rest of the comments should also be addressed.

Slartibartfass2 reviewed (changes requested):
> I could only confirm one missing instance of the
> 
> ```
> @DelicateCoroutinesApi
> @ExperimentalCoroutinesApi
> ```
> 
> annotations. The rest were definitely removed previously. The review system may have showed you an older version. Either way, there is no trace of them in the source code anymore. The rest of the comments should also be addressed.

I only annotated this once, but there are some other open conversations left at the same place.
You can rebase now.

luca-schlecker commented:
Rebased and squashed commits :+1:

Slartibartfass2 reviewed (commented):
There are open unanswered conversations left. After answering them, we can merge this PR.

luca-schlecker commented:
All of them have already been fixed in the source code.

Slartibartfass2 reviewed (commented):
Yes, they weren't answered after the previous reviews, so please answer them now or at least give them a thumbs up.

luca-schlecker commented:
> Yes, they weren't answered after the previous reviews, so please answer them now or at least give them a thumbs up.

This is unnecessary. Just resolve them already.

Slartibartfass2 reviewed (approved):
Please propose changes to our conventions if you think they are unnecessary. We all agreed to adhere to them.

luca-schlecker commented:
I neither can remember agreeing to react to every single comment nor could I find any official document stating this convention. Do you have any lead to where this was written down?

Slartibartfass2 commented:
> I neither can remember agreeing to react to every single comment nor could I find any official document stating this convention. Do you have any lead to where this was written down?

You can find the contribution guidelines in the snowballr repo wiki. Previously they were in the frontend repo.

There was a proposal PR and everyone in our team channel was asked to have a look at it. It was merged after 11 days with no objections.

luca-schlecker commented:
> > I neither can remember agreeing to react to every single comment nor could I find any official document stating this convention. Do you have any lead to where this was written down?
> 
> You can find the contribution guidelines in the snowballr repo wiki. Previously they were in the frontend repo.
> 
> There was a proposal PR and everyone in our team channel was asked to have a look at it. It was merged after 11 days with no objections.

I fully missed out on the existence of this document and its creation. Thank you for guiding me there.
I do think that this guideline is good, I even (mostly) adopted it without knowing of its existence, just by pure intuition. I also value the efficiency-gains by streamlining the review process. However, I do not value enforcing a guideline that has already failed. There were comments left unanswered, you rightfully brought that up and I continued that conversation. The original issue is already solved: knowing what has been addressed. From my point of view, there is no benefit in requiring that thumbs-up, just for it to be there.
Would you be open to relaxing this hard requirement? I'd like to keep this guideline in place, this should be our standard way of doing things. But if the goal of that guideline has already been achieved, there is no need in enforcing it.

Slartibartfass2 commented:
> > > I neither can remember agreeing to react to every single comment nor could I find any official document stating this convention. Do you have any lead to where this was written down?
> > 
> > You can find the contribution guidelines in the snowballr repo wiki. Previously they were in the frontend repo.
> > 
> > There was a proposal PR and everyone in our team channel was asked to have a look at it. It was merged after 11 days with no objections.
> 
> I fully missed out on the existence of this document and its creation. Thank you for guiding me there.
> I do think that this guideline is good, I even (mostly) adopted it without knowing of its existence, just by pure intuition. I also value the efficiency-gains by streamlining the review process. However, I do not value enforcing a guideline that has already failed. There were comments left unanswered, you rightfully brought that up and I continued that conversation. The original issue is already solved: knowing what has been addressed. From my point of view, there is no benefit in requiring that thumbs-up, just for it to be there.
> Would you be open to relaxing this hard requirement? I'd like to keep this guideline in place, this should be our standard way of doing things. But if the goal of that guideline has already been achieved, there is no need in enforcing it.
> 

Where do we draw the line then? One could just implement the changes but never would need to answer any conversation.

luca-schlecker commented:
> > > > I neither can remember agreeing to react to every single comment nor could I find any official document stating this convention. Do you have any lead to where this was written down?
> > > 
> > > 
> > > You can find the contribution guidelines in the snowballr repo wiki. Previously they were in the frontend repo.
> > > There was a proposal PR and everyone in our team channel was asked to have a look at it. It was merged after 11 days with no objections.
> > 
> > 
> > I fully missed out on the existence of this document and its creation. Thank you for guiding me there.
> > I do think that this guideline is good, I even (mostly) adopted it without knowing of its existence, just by pure intuition. I also value the efficiency-gains by streamlining the review process. However, I do not value enforcing a guideline that has already failed. There were comments left unanswered, you rightfully brought that up and I continued that conversation. The original issue is already solved: knowing what has been addressed. From my point of view, there is no benefit in requiring that thumbs-up, just for it to be there.
> > Would you be open to relaxing this hard requirement? I'd like to keep this guideline in place, this should be our standard way of doing things. But if the goal of that guideline has already been achieved, there is no need in enforcing it.
> 
> Where do we draw the line then? One could just implement the changes but never would need to answer any conversation.

I don't think there needs to be a clear line to be drawn. It stays the default and preferred way of handling things, but does not have to be a reason for blocking an otherwise totally fine PR.
